aerosol_dynamics/particle.py

```diff
@@ -0,0 +1,184 @@
+            radius (float): The radius of the particle.
+            density (float): The density of the particle.
+            charge (float): The charge of the particle.
+        """
+        self._name = name
+        self._radius = radius
+        # self._density = density
+        self._charge = charge
+        self._mass = density * (4*np.pi/3) * (radius**3)
+
+    def name(self) -> str:
+        """Returns the name of the particle."""
+        return self._name
+
+    @u.wraps(u.kg, [None])
+    def mass(self) -> float:
+        """Returns mass of a particle. Checks units. [kg]"""
+        return self._mass
+
+    @u.wraps(u.m, [None])
+    def radius(self) -> float:
+        """Returns radius of a particle. Checks units. [m]"""
+        return self._radius
+
+    @u.wraps(u.dimensionless, [None])
+    def charge(self) -> float:
+        """Returns charge of a particle. Checks units. [unitless]"""
+        return self._charge
+
+    @u.wraps(u.dimensionless, [None])
+    def knudsen_number(self) -> float:
+        """Returns a particle's Knudsen number. Unitless.
+
+        The Knudsen number seeks to reflect the relative length scales of the
+        particle and the suspending fluid (air, water, etc.). This is calculated
+        by the mean free path of the medium divided by the particle radius.
+        """
+        return pp.MEAN_FREE_PATH_AIR / self.radius()
+
+    @u.wraps(u.dimensionless, [None])
+    def slip_correction_factor(self) -> float:
+        """Returns a particle's Cunningham slip correction factor. Unitless.
+
+        The slip correction factor is a dimensionless quantity that accounts for
+        non-continuum effects when calculating the drag on small particles.
+        This is a deviation from Stokes' Law; Stokes assumes a no-slip
+        condition that is no longer correct at high Knudsen numbers.
+        The slip correction factor is used to calculate the friction factor.
+        See Eq 9.34 in Atmos. Chem. & Phys. (2016) for more information.
+        """
+        knudsen_number: float = self.knudsen_number()
+        return 1 + knudsen_number * (1.257 + 0.4*np.exp(-1.1/knudsen_number))
+
+    @u.wraps(u.kg / u.s, [None])
+    def friction_factor(self) -> float:
+        """Returns a particle's friction factor. [N-s/m].
+
+        The friction factor is a property of the particle's size and the medium
+        that the particle is in. Multiplying the friction factor by the fluid
+        velocity gives the drag force on the particle.
+        """
+        slip_correction_factor: float = self.slip_correction_factor()
+        return 6 * np.pi * pp.MEDIUM_VISCOSITY * self.radius() / slip_correction_factor
+
+    @u.wraps(u.kg, [None, None])
+    def reduced_mass(self, other) -> float:
+        """Returns the reduced mass of two particles. [kg].
+
+        The reduced mass is an "effective inertial" mass that allows a two body
+        problem to be solved as a one body problem.
```

Comment:
what does that two body to one body mean? (in our context) I think we'll have to explain this in an example eventually.

Reply:
If you look here, you will see we are not using the two masses, instead just the reduced mass. Maybe a more practical way to look at this would be to consider the reduced mass as a correction of the size effects on the second line.  

https://github.com/uncscode/particula/blob/3c978f7ae83d166cb12fd6f63526e37744d115fb/aerosol_dynamics/particle.py#L131-L132

Reply:
Btw, I would consider rewriting this example like this in the future (no need to rewrite it now --- we will likely need to rewrite it later anyway for corrections; I think I have made a mistake or two somewhere in the calculations that the tests will later help us figure out; this calculation doesn't exactly coincide with the values I have in mind)

```python
    return (
        (
            pp.TEMPERATURE*pp.BOLTZMANN_CONSTANT*reduced_mass**0.5
        )*coulomb_enhancement_continuum_limit/(
            reduced_friction_factor*(self.radius() + other.radius())
        )*coulomb_enhancement_kinetic_limit
    ) 
```

Reply:
Okay, just a little more context on the "reduced mass" thingy:

Usually you could get away with assuming the smaller particle is mobile while the bigger one is stationary, thus a coagulation problem is really just a slightly modified condensation problem or vice versa. And that actually works remarkably well, except when the particles have similar sizes (or volumes or masses) and especially in the free molecular regime (i.e. both particles are small enough compared to the surrounding gas). Essentially, that's really what's hard about coagulation --- you've got to link the two relatively well-understood limits through the transition regime (both in terms of particle--gas interactions as well as particle--particle interactions). So you start taking into account all higher-order dependencies because they actually matter. The reduced mass definition goes like (m1 x m2)/(m1 + m2), let's say m1 > m2, and remember that people often treat the smaller particle as mobile and the bigger particle as stationary, then you have usually just take the mass in question as m2 (the smaller mass). The reduced mass is just a correction on that. If m1 >> m2, then (m1 x m2)/(m1 + m2) = 1 (approx); that is, the correction is negligible. But if the masses are similar, then, this correction can be significant (the correction is always equal to or less than 1: m1/(m1+m2) <= 1; hence "reduced" in reduced mass, the mass of the smaller particle is effectively reduced). 

What does this actually mean? It just means both are moving and both are contributing to the collision --- remember...

Reply:
Just me venting about how complex equations don't translate efficiently into coding style: 
Making long equations readable is something I'm still trying to get comfortable with. I tend to read equations horizontally, so parsing equations vertically make it harder for me to get the big picture. Similarly, splitting open and close parentheses on different lines breaks the visual "this is a group" that parentheses are supposed to give. In addition, in the example I would want the division sign / should be more prominent, because it effectively splits the expression in two parts. In perhaps conventional coding style, that is not as clear. I understand the desire to limit # characters in a line, but I feel it can be restrictive at times and not allow for capturing complex equations well. 

If we have to rewrite it to fit our style guide though, then we'll rewrite it.

Reply:
Adding an example is a good idea @Gorkowski! I only added a simple explanation that I pulled from wiki, but it's worth painting a clearer picture in our documentation. My hope is that a first year grad student would be able to go through this and understand almost everything on a surface level without having to refer to outside texts or googling (and perhaps some forgetful phd students too), though I'm open to having more or less explanations.

Reply:
> splitting open and close parentheses on different lines breaks the visual "this is a group" that parentheses are supposed to give

quite the opposite actually, no? Which one of these asserts the groupness of expressions between the parentheses more prominently?

```python
(
    five*time*twnetytwo + fifty*times*twothousand
)/(
    something_else
)
```

```python
(five*time*twnetytwo + fifty*times*twothousand)/
(something_else)
```

```python
(five*time*twnetytwo + fifty*times*twothousand)/(something_else)
```

I don't think what you're doing is wrong --- it is just my opinion that writing equations vertically makes it easier to debug them. And this happens to get me around the line limits and all that stuff imposed by the linter.

Reply:
I think it's just what my brain is used to, which does more visual grouping. The separating of parentheses and extra lines is like extra whitespace to me -- distracting (unless it serves to separate nested parentheses, which are visually confusing!). There's sort of a grey area where we transition from horizontal to vertical, right? -- we probably wouldn't split something short like:
(
    argone + argtwo
) * argthree
because it's easier to read as: (argone+argtwo) * argthree. (To me at least!)

And I would prefer:
(first group) * factor_one / 
(
    (
        (second group) + 
        (third group)
    ) * factor_two
)
over:
(
    first group
) * factor_one / 
(
    (
        (
            second group
        ) + 
        (
            third group
        )
    ) * factor_two
)

BUT, I am just elaborating on my preferences, which is amenable to change and adaptation! Our style guide is silent on it, and it appears to have no clear consensus online, so I think it comes down to what we are comfortable with and what communicates well. I don't want to be strictly bound to an 80 character line limit because a linter says so (and wasn't developed with math equations in mind) when we have descriptive variable names. Sure it's good to try to keep it shorter most of the time, but splitting too much vertically can be more difficult to understand too. I think I sound more stubborn than I feel, so I want to make it clear that I'm open to a new style. Who knows, as we write more equations, I might find that my style is not great for some of them. Any thoughts or preferences, @Gorkowski?

Reply:
I would say the following on the 80-col (or 72-col) limit: If you can follow it, it is really good and make the code more portable (i.e. easier to view in multiple editors like vim, vscode, github.dev, etc.)

But I wouldn't go out of my way to break a long link jus to satisfy. However, I would break a long path (../.../..../..../..../...) to satisfy it. I have recently started reading some old-school programming stuff and I am influenced by them (the unix philosophy, anyone? _awk & sed_ has been fun!)

The idea here is not to be stubborn and picky, and I would really hate for this stuff to delay us. If you give me your code with no formatting and it works, I am happy with it and I can volunteer to lint it --- or just leave it as is. At the end of the day, I don't want any of these random extras to derail our progress!

Reply:
And I think what matters most is 1) consistency within one document and 2) consistency in code overall --- in this order.

Reply:
Fair enough! Well, I think I'm happy to switch it to a more vertical profile. I question how much old-school stuff still applies (e.g. monitors are wider these days, though I realize laptop screens are still small). But you are absolutely right, I wouldn't want this to prevent progress on the more substantial areas either. Refactoring is *always* an option later. I'll change some of these expressions and push another version tonight.

Reply:
> I'll change some of these expressions and push another version tonight.

Please don't. Let's worry about that later. This is more than good enough for now. We can revisit. And don't worry, I am outspoken and I bring things up, so you will hear more from me :)